MySQL source: Add comprehensive data type test

DoNotPanicUA opened the pull request:
Issue: #3562 

## What
We are creating comprehensive integration tests for the MySQL source connector. Such a test guarantees that connector implementation is infallible even in cases when the source contains unexpected/extreme values.
In addition, we provide an abstract implementation of the comprehensive test `SourceComprehensiveTest` in order to simplify test creation.

## How
1. Prepare a list of cases for testing in issue #3562.
2. Implement an abstract comprehensive test in order to simplify the implementation. A common part for the existing acceptance test and the created comprehensive test is moved to a common class `SourceAbstractTest.java`.
3. Implement the abstract test with all cases from the issue.

## Pre-merge Checklist
- [ ] *Run integration tests*

## Recommended reading order
1. `MySqlSourceComprehensiveTest.java`, `CdcMySqlSourceComprehensiveTest.java`
2. `SourceComprehensiveTest.java`, `SourceAbstractTest.java`
3. the rest

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+  private static final String DEFAULT_INSERT_SQL = "INSERT INTO %1$s VALUES (%2$s, %3$s);";
+
+  private final String sourceType;
+  private final Field.JsonSchemaPrimitive airbyteType;
```
@DoNotPanicUA please move jsonSchema to a separate class.

yaroslav-hrytsaienko replied:
@DoNotPanicUA please move jsonSchema to a separate class.

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class DataTypeTest {
```
@DoNotPanicUA I suggest to change the name of the class. Basically it is not a test it self. Maybe something like test data wrapper, or smth like this would be better.

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+    return airbyteType;
+  }
+
+  public String getName() {
```
@DoNotPanicUA please rename the method to getNameWithTestPrefix

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+    return "test_" + testNumber + "_" + sourceType;
+  }
+
+  public String getCreateSQL() {
```
@DoNotPanicUA getCreateSQL-> getCreateSqlQuery

yaroslav-hrytsaienko replied:
same for others

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+     * @param insertValue test value
+     * @return builder
+     */
+    public DataTypeTestBuilder addInsertValue(String... insertValue) {
```
@DoNotPanicUA addInsertValue->addInsertValues

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/DataTypeTest.java:
```diff
@@ -0,0 +1,179 @@
+     * @param createTablePatternSQL creation table sql pattern
+     * @return builder
+     */
+    public DataTypeTestBuilder createTablePatternSQL(String createTablePatternSQL) {
```
@DoNotPanicUA due to JCC createTablePatternSQL->createTablePatternSql

DoNotPanicUA replied:
fixed everywhere

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/SourceAbstractTest.java:
```diff
@@ -0,0 +1,158 @@
+   * @param testEnv - information about the test environment.
+   * @throws Exception - can throw any exception, test framework will handle.
+   */
+  protected abstract void setup(TestDestinationEnv testEnv) throws Exception;
```
@DoNotPanicUA  setup-> setupEnvironment , testEnv-> environment

yaroslav-hrytsaienko replied:
same for others

yaroslav-hrytsaienko commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/SourceComprehensiveTest.java:
```diff
@@ -0,0 +1,147 @@
+   *
+   * @param test comprehensive data type test
+   */
+  public void addDataTypeTest(DataTypeTest test) {
```
@DoNotPanicUA  addDataTypeTest -> addDataTypeTestData

yaroslav-hrytsaienko commented on airbyte-integrations/connectors/source-mysql/src/test-integration/java/io/airbyte/integrations/source/mysql/MySqlSourceComprehensiveTest.java:
```diff
@@ -0,0 +1,248 @@
+
+  @Override
+  protected void initTests() {
+    addDataTypeTest(
```
@DoNotPanicUA in case of datasets are equal please move them to parent class

DoNotPanicUA replied:
they are slight different.

yaroslav-hrytsaienko reviewed (commented):
@DoNotPanicUA  great work! Needs a bit of tuning.

subodh1810 commented on airbyte-integrations/bases/standard-source-test/src/main/java/io/airbyte/integrations/standardtest/source/SourceComprehensiveTest.java:
```diff
@@ -0,0 +1,147 @@
+    final List<AirbyteMessage> recordMessages = allMessages.stream().filter(m -> m.getType() == Type.RECORD).collect(Collectors.toList());
+
+    recordMessages.forEach(msg -> LOGGER.debug(msg.toString()));
+    assertFalse(recordMessages.isEmpty(), "Expected records from source");
```
I think we should assert proper values instead of just values being present

DoNotPanicUA replied:
Done

subodh1810 reviewed (changes requested):
Few changes : 
1. We should split the tests. Each data type should have its own individual tests
2. There should be individual assertion for each data type. We should check the value is correct and not just that the value is present.
3. We should try to build the tests in such a way that they can be extensible for different JDBC connectors in the future

DoNotPanicUA commented:
> Few changes :
> 
> 1. We should split the tests. Each data type should have its own individual tests
> 2. There should be individual assertion for each data type. We should check the value is correct and not just that the value is present.
> 3. We should try to build the tests in such a way that they can be extensible for different JDBC connectors in the future

Hi, @subodh1810.
Thank you for your review.
The first implementation was focused on testing that the streamer is infallible even if source data is quite "tricky". But I agree it's not a "comprehensive" test if we don't check a streamer result.
So, I've added the possibility to specify expected values if it's required and corresponding validations.
In addition, I've added all found bugs (from my perspective) to [the table](https://github.com/airbytehq/airbyte/issues/3562#issuecomment-850838160).

About your third point. I realize that big difference between JDBC sources and I don't see common data tests for all JDBC sources. So, I decided to create the builder which allows you to describe your tests for any JDBC source. All common operations I moved to the abstract class.
If you have some ideas on how to build common tests for all JDBC sources, please call me at any convenient time for you.

subodh1810 commented on airbyte-integrations/connectors/source-mysql/src/test-integration/java/io/airbyte/integrations/source/mysql/CdcMySqlSourceComprehensiveTest.java:
```diff
@@ -0,0 +1,372 @@
+            .airbyteType(JsonSchemaPrimitive.STRING)
+            .addInsertValues("null", "1", "127", "-128")
+            // @TODO returns number instead of boolean
+            // .addExpectedValues(null, "true", "false", "false")
```
This should be fixed by the PR https://github.com/airbytehq/airbyte/pull/3890

subodh1810 reviewed (commented):
@DoNotPanicUA can you run the tests for jdbc sources using `/test connector=source-mysql`

subodh1810 commented:
/test connector=source-mysql

subodh1810 reviewed (approved):
Can you create a follow up task to fix the TODOs. Also please run the tests for all the JDBC source connectors using the github command `/test connector` to make sure the change would not cause any problem.  Rest looks good.

DoNotPanicUA commented:
/test connector=source-jdbc

DoNotPanicUA commented:
/test connector=source-mssql

DoNotPanicUA commented:
/test connector=source-oracle

DoNotPanicUA commented:
/test connector=source-redshift

DoNotPanicUA commented:
/test connector=source-clickhouse
